lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -285,11 +285,18 @@ static LogicalResult canonicalizePrimOp(
   SmallVector<Attribute, 3> constOperands;
   constOperands.reserve(op->getNumOperands());
   for (auto operand : op->getOperands()) {
-    Attribute attr;
-    if (auto *defOp = operand.getDefiningOp())
-      TypeSwitch<Operation *>(defOp).Case<ConstantOp, SpecialConstantOp>(
-          [&](auto op) { attr = op.fold({}).template get<Attribute>(); });
-    constOperands.push_back(attr);
+    if (auto constantOp = operand.getDefiningOp<ConstantOp>()) {
+      auto cst = constantOp.fold(ConstantOp::FoldAdaptor({})).get<Attribute>();
+      constOperands.push_back(cst);
```

Comment:
Maybe we can use `getValueAttr` normally?
```suggestion
      constOperands.push_back(constantOp.getValueAttr());
```

Reply:
Great suggestion! thanks